MVP implementation

YuriyVaadin opened the pull request:
Connects to https://github.com/vaadin/vaadin-core/issues/136

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/vaadin/vaadin-dropdown-menu/2)
<!-- Reviewable:end -->

abdonrd commented on vaadin-dropdown-menu.html:
```diff
@@ -6,48 +6,200 @@
-      class VaadinDropdownMenuElement extends Vaadin.ThemableMixin(Polymer.Element) {
-        static get is() {
-          return 'vaadin-dropdown-menu';
+  <dom-module id="vaadin-dropdown-menu-overlay-styles" theme-for="vaadin-dropdown-menu-overlay">
```
Extra indent.

abdonrd commented on vaadin-dropdown-menu.html:
```diff
@@ -14,40 +64,142 @@
+        */
       window.Vaadin = window.Vaadin || {};
-      Vaadin.VaadinDropdownMenuElement = VaadinDropdownMenuElement;
+      Vaadin.DropdownMenuElement = VaadinDropdownMenuElement;
```
As other vaadin elements, I think you should change `VaadinDropdownMenuElement ` to `DropdownMenuElement`.

abdonrd replied:
@manolo check:

`vaadin-tabs` & `vaadin-tab`:
https://github.com/vaadin/vaadin-tabs/blob/12bd28a6cd924a3caf3f9f4c97f4bc89bf878368/vaadin-tabs.html#L235-L236

https://github.com/vaadin/vaadin-tabs/blob/12bd28a6cd924a3caf3f9f4c97f4bc89bf878368/vaadin-tab.html#L102-L103

`vaadin-button`:
https://github.com/vaadin/vaadin-button/blob/a4972d26c8934fcfbf99b56ede420d76cc03a86a/vaadin-button.html#L108-L109

`vaadin-radio-button`:
https://github.com/vaadin/vaadin-radio-button/blob/4a250e6b8b2cada6e2338be1d143669860a8d7f6/vaadin-radio-button.html#L201-L202

abdonrd replied:
Dont worry, thanks Manolo!

manolo commented:
Review status: 0 of 12 files reviewed at latest revision, 2 unresolved discussions.

---

*[vaadin-dropdown-menu.html, line 202 at r4](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-KwnbpaByHRphOXfWLZe-r4-202:-KwoR32SEXYuVbLf6rE6:b7f8dw2) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/af32374bf3f3d5dc37ac45cf778f792418fb5cf4/vaadin-dropdown-menu.html#L202)):*
<details><summary><i>Previously, abdonrd (Abdón Rodríguez Davila) wrote…</i></summary><blockquote>

As other vaadin elements, I think you should change `VaadinDropdownMenuElement ` to `DropdownMenuElement`.
</blockquote></details>

In all our elements, we use the full `VaadinDropdownMenuElement` for the variable in the element definition scoped block (note that the variable is not visible out of the block), but we remove the `Vaadin` prefix  when making the class public in `window.Vaadin`  namespace

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2)*
<!-- Sent from Reviewable.io -->

manolo commented:
Review status: 0 of 12 files reviewed at latest revision, 3 unresolved discussions.

---

*[bower.json, line 41 at r5](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-KwoS2-O0hHwLclLpUl0:-KwoS2-O0hHwLclLpUl1:b-fb9ycn) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/c0fc8d3e70aa2623020e4a4aeda15d8bd2d8ecd6/bower.json#L41)):*
> ```JSON
>   },
>   "resolutions": {
>     "vaadin-list-mixin": "^1.1.1"
> ```

Use 1.2.0

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2)*
<!-- Sent from Reviewable.io -->

YuriyVaadin commented:
Review status: 0 of 12 files reviewed at latest revision, 3 unresolved discussions.

---

*[bower.json, line 41 at r5](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-KwoS2-O0hHwLclLpUl0:-Kwo_KEfUg4B2d5ceeq3:b-896fix) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/c0fc8d3e70aa2623020e4a4aeda15d8bd2d8ecd6/bower.json#L41)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

Use 1.2.0
</blockquote></details>

Done.

---

*[vaadin-dropdown-menu.html, line 12 at r3](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-KwnbpaByHRphOXfWLZe-r3-12:-KwoZeB6sxMG0ofABov9:bss4a2b) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/77d6aefe3058702398a0011a8f0dfc23cdba1653/vaadin-dropdown-menu.html#L12)):*
<details><summary><i>Previously, abdonrd (Abdón Rodríguez Davila) wrote…</i></summary><blockquote>

Extra indent.
</blockquote></details>

Removed.

---

*[vaadin-dropdown-menu.html, line 202 at r4](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-KwnbpaByHRphOXfWLZe-r4-202:-Kwo_HAPhYxYVXOTkWEk:b-a2ujfl) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/af32374bf3f3d5dc37ac45cf778f792418fb5cf4/vaadin-dropdown-menu.html#L202)):*
<details><summary><i>Previously, abdonrd (Abdón Rodríguez Davila) wrote…</i></summary><blockquote>

@manolo check:

`vaadin-tabs` & `vaadin-tab`:
https://github.com/vaadin/vaadin-tabs/blob/12bd28a6cd924a3caf3f9f4c97f4bc89bf878368/vaadin-tabs.html#L235-L236

https://github.com/vaadin/vaadin-tabs/blob/12bd28a6cd924a3caf3f9f4c97f4bc89bf878368/vaadin-tab.html#L102-L103

`vaadin-button`:
https://github.com/vaadin/vaadin-button/blob/a4972d26c8934fcfbf99b56ede420d76cc03a86a/vaadin-button.html#L108-L109

`vaadin-radio-button`:
https://github.com/vaadin/vaadin-radio-button/blob/4a250e6b8b2cada6e2338be1d143669860a8d7f6/vaadin-radio-button.html#L201-L202
</blockquote></details>

Changed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2)*
<!-- Sent from Reviewable.io -->

manolo commented:
Review status: 0 of 12 files reviewed at latest revision, 3 unresolved discussions.

---

*[vaadin-dropdown-menu.html, line 202 at r4](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-KwnbpaByHRphOXfWLZe-r4-202:-KwocsOi9VA2MKAPM4tO:b-ysa0dp) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/af32374bf3f3d5dc37ac45cf778f792418fb5cf4/vaadin-dropdown-menu.html#L202)):*
<details><summary><i>Previously, YuriyVaadin (Yuriy Yevstihnyeyev) wrote…</i></summary><blockquote>

Changed.
</blockquote></details>

Good point Abdón.
I was confused because of the skeleton, so there are still two recent components that I defined wrong: item and listbox, I will change soon.
Thanks

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2)*
<!-- Sent from Reviewable.io -->

tomivirkki commented:
*[vaadin-dropdown-menu.html, line 80 at r6](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-Kwoja0OW6hEglF7Agsh:-Kwoja0OW6hEglF7Agsi:bh6dm2e) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/9f64c35df6e544f4d11665695bd0d6aa1ca48dc8/vaadin-dropdown-menu.html#L80)):*
> ```HTML
>         * ```
>         * <vaadin-dropdown-menu>
>         *  <template>
> ```

Add listbox to this example

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2)*
<!-- Sent from Reviewable.io -->

manolo commented:
Review status: 0 of 12 files reviewed at latest revision, 17 unresolved discussions.

---

*[vaadin-dropdown-menu.html, line 46 at r6](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-KwohGZC37Wjwdk1danP:-KwohGZC37Wjwdk1danQ:btmvq47) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/9f64c35df6e544f4d11665695bd0d6aa1ca48dc8/vaadin-dropdown-menu.html#L46)):*
> ```HTML
>         *
>         * See [`<vaadin-overlay>` documentation](https://github.com/vaadin/vaadin-overlay/blob/master/vaadin-overlay.html)
>         * for `<vaadin-dropdown-menu-overlay>` parts.
> ```

I think we need to define here the parts that `Vaadin.OverlayElement` has

---

*[vaadin-dropdown-menu.html, line 68 at r6](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-KwozmAz3V_nGChxxrL0:-KwozmAz3V_nGChxxrL1:b186vll) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/9f64c35df6e544f4d11665695bd0d6aa1ca48dc8/vaadin-dropdown-menu.html#L68)):*
> ```HTML
>     </style>
> 
>     <vaadin-text-field readonly value="[[value]]" id="input"></vaadin-text-field>
> ```

If we want to be iron-form compatible we need to forward at least name and validate, but also disabled ...

---

*[vaadin-dropdown-menu.html, line 69 at r6](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-KwozH-lqy7qx0DYsHQv:-KwozH-lqy7qx0DYsHQw:b-2ewlue) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/9f64c35df6e544f4d11665695bd0d6aa1ca48dc8/vaadin-dropdown-menu.html#L69)):*
> ```HTML
> 
>     <vaadin-text-field readonly value="[[value]]" id="input"></vaadin-text-field>
>     <vaadin-dropdown-menu-overlay opened="{{opened}}" id="overlay">
> ```

Nit: if there is no content, I prefer in one line like the line above. Or anyways align both with either one or two lines

---

*[vaadin-dropdown-menu.html, line 101 at r6](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-KwsIIXENzMqqPDeDuwr:-KwsIIXENzMqqPDeDuws:bpifbew) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/9f64c35df6e544f4d11665695bd0d6aa1ca48dc8/vaadin-dropdown-menu.html#L101)):*
> ```HTML
>             },
> 
>             _contentTemplate: Object,
> ```

Why do we need this property?. It's used once in the same block

---

*[vaadin-dropdown-menu.html, line 124 at r6](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-Kwp-MnPpj_DbbCJblv8:-Kwp-MnPpj_DbbCJblv9:buhkb46) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/9f64c35df6e544f4d11665695bd0d6aa1ca48dc8/vaadin-dropdown-menu.html#L124)):*
> ```HTML
> 
>         _onKeyDown(e) {
>           if (!this.opened && /^(ArrowDown|Down|ArrowUp|Up|Enter|Space)$/.test(e.code)) {
> ```

Native select does not open with enter, but paper-drop-down does. I feel that we should fill native behavior but not sure @limonte @platosha  ?

---

*[vaadin-dropdown-menu.html, line 131 at r6](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-Kwp09A-MW0QFEvFv54G:-Kwp09A-MW0QFEvFv54H:b-jw8p36) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/9f64c35df6e544f4d11665695bd0d6aa1ca48dc8/vaadin-dropdown-menu.html#L131)):*
> ```HTML
> 
>         _onKeyDownInside(e) {
>           const menu = this.$.overlay.$.content.lastElementChild;
> ```

1. we are avoiding to use ids because of avoiding conflicting id in shady mode, @tomivirkki  wdyt, should we remove the id here as well?
2. I don't think lastElementChild is the safest way to get the menu element, I would check for elements that has the mixin, or the localName
3. btw there is a overlayElement getter in this class, use it.

---

*[vaadin-dropdown-menu.html, line 143 at r6](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-KwsGhKT9gbEn27VImHF:-KwsGhKT9gbEn27VImHG:bf0qpp3) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/9f64c35df6e544f4d11665695bd0d6aa1ca48dc8/vaadin-dropdown-menu.html#L143)):*
> ```HTML
>           };
> 
>           if ((e.type === 'keydown' && /^(Enter|Space)$/.test(e.code) || e.type === 'click') && eventFromSelected) {
> ```

should use `e.key` instead
move boolean `evenFromSelected` at the beginning of the condition, checking booleans first make the condition perform better.

---

*[vaadin-dropdown-menu.html, line 146 at r6](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-KwsH6AT-76pCU2A6D8o:-KwsH6AT-76pCU2A6D8p:bqf8zgd) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/9f64c35df6e544f4d11665695bd0d6aa1ca48dc8/vaadin-dropdown-menu.html#L146)):*
> ```HTML
>             if (this.value === '') {
>               this.value = e.composedPath()[0].textContent;
>             }
> ```

more concise `this.value === '' && this.value = e. ... ` or `this.value = this.value || ...`

---

*[vaadin-dropdown-menu.html, line 148 at r6](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-KwsHyoL7xPyUaJO9FXO:-KwsHyoL7xPyUaJO9FXP:b1ivbet) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/9f64c35df6e544f4d11665695bd0d6aa1ca48dc8/vaadin-dropdown-menu.html#L148)):*
> ```HTML
>             }
>             this.opened = false;
>           } else if (e.type === 'keydown' && /^(Tab)$/.test(e.code)) {
> ```

`e.key`

---

*[vaadin-dropdown-menu.html, line 156 at r6](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-KwsIRgG3l3Gx951zREz:-KwsIRgG3l3Gx951zRF-:b345hhd) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/9f64c35df6e544f4d11665695bd0d6aa1ca48dc8/vaadin-dropdown-menu.html#L156)):*
> ```HTML
>         _openedChanged(opened) {
>           if (opened && !this._instance) {
>             this._contentTemplate = this.querySelector('template');
> ```

Seems that it should be a `const _contentTemplate` instead of an instance property. It's not used anywhere else

---

*[vaadin-dropdown-menu.html, line 174 at r6](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-KwsIiSWxd5polH7PjmM:-KwsIiSWxd5polH7PjmN:b-oczwf5) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/9f64c35df6e544f4d11665695bd0d6aa1ca48dc8/vaadin-dropdown-menu.html#L174)):*
> ```HTML
>           this.overlayElement.opened = opened;
> 
>           const menu = this.$.overlay.$.content.lastElementChild;
> ```

The same here. I guess menu could be computed once and leave it in a instance variable

---

*[vaadin-dropdown-menu.html, line 187 at r6](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-KwpuYHZ2z8VG7xZHboE:-KwpuYH_kUzJmS4js1Hq:b-12pljc) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/9f64c35df6e544f4d11665695bd0d6aa1ca48dc8/vaadin-dropdown-menu.html#L187)):*
> ```HTML
>             }
> 
>             menu.addEventListener('selected-changed', selectedChangedListener);
> ```

what's the reason of adding/removing 

---

*[vaadin-dropdown-menu.html, line 199 at r6](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-KwsJ4O7fMD7l39jaOSU:-KwsJ4O7fMD7l39jaOSV:bizsd2k) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/9f64c35df6e544f4d11665695bd0d6aa1ca48dc8/vaadin-dropdown-menu.html#L199)):*
> ```HTML
>               menu.removeEventListener('selected-changed', selectedChangedListener);
>               menu.removeEventListener('keydown', this._onKeyDownInside.bind(this));
>               menu.removeEventListener('click', this._onKeyDownInside.bind(this));
> ```

I feel like we should compute the menu element once when the light dom is available and add the events, I dont see the need of adding/removing events each time it's opened/closed

---

*[vaadin-dropdown-menu.html, line 215 at r6](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-KwsJY1twG-Aq9_2kc-i:-KwsJY1twG-Aq9_2kc-j:b-24pynq) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/9f64c35df6e544f4d11665695bd0d6aa1ca48dc8/vaadin-dropdown-menu.html#L215)):*
> ```HTML
> 
>         get overlayElement() {
>           return this.$.overlay;
> ```

@tomivirkki shouldn't we avoid ids ? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2)*
<!-- Sent from Reviewable.io -->

manolo commented:
Review status: 0 of 12 files reviewed at latest revision, 19 unresolved discussions.

---

*[demo/dropdown-menu-basic-demos.html, line 11 at r6](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-KwsKfPT8BkYOv-UpaUj:-KwsKfPT8BkYOv-UpaUk:bcms946) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/9f64c35df6e544f4d11665695bd0d6aa1ca48dc8/demo/dropdown-menu-basic-demos.html#L11)):*
> ```HTML
> 
> 
>     <h3>Sample example with vaadin-list-box</h3>
> ```

Don't we only support list-boxes as content? hence title should be just `Basic Usage` or `Usage`

---

*[test/dropdown-menu-test.html, line 35 at r6](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-KwsL8FcceBCcwVkVKJ-:-KwsL8FcceBCcwVkVKJ0:b-dmh5xj) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/9f64c35df6e544f4d11665695bd0d6aa1ca48dc8/test/dropdown-menu-test.html#L35)):*
> ```HTML
>       it('should have a valid localName', function() {
>         expect(dropdown.localName).to.be.equal('vaadin-dropdown-menu');
>       });
> ```

we need more tests here, like testing selecting, focus code, etc

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2)*
<!-- Sent from Reviewable.io -->

YuriyVaadin commented:
Review status: 0 of 12 files reviewed at latest revision, 19 unresolved discussions.

---

*[vaadin-dropdown-menu.html, line 46 at r6](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-KwohGZC37Wjwdk1danP:-Kwsi_HvbYO1Vv5DbWJE:bqmjwcz) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/9f64c35df6e544f4d11665695bd0d6aa1ca48dc8/vaadin-dropdown-menu.html#L46)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

I think we need to define here the parts that `Vaadin.OverlayElement` has
</blockquote></details>

Added.

---

*[vaadin-dropdown-menu.html, line 68 at r6](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-KwozmAz3V_nGChxxrL0:-KwsiaNwQXHvL2b6z2rY:bwa86r6) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/9f64c35df6e544f4d11665695bd0d6aa1ca48dc8/vaadin-dropdown-menu.html#L68)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

If we want to be iron-form compatible we need to forward at least name and validate, but also disabled ...
</blockquote></details>

`name` and `disabled` were added. Should `validate` method be the same as in `radio-button-group` for example? https://github.com/vaadin/vaadin-radio-button/blob/master/vaadin-radio-button-group.html#L221

---

*[vaadin-dropdown-menu.html, line 69 at r6](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-KwozH-lqy7qx0DYsHQv:-Kwsj0xq4tWwNuX3kRkc:b-a2ujfl) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/9f64c35df6e544f4d11665695bd0d6aa1ca48dc8/vaadin-dropdown-menu.html#L69)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

Nit: if there is no content, I prefer in one line like the line above. Or anyways align both with either one or two lines
</blockquote></details>

Changed.

---

*[vaadin-dropdown-menu.html, line 80 at r6](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-Kwoja0OW6hEglF7Agsh:-Kwsj25sN_STSjkfH8i6:bqmjwcz) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/9f64c35df6e544f4d11665695bd0d6aa1ca48dc8/vaadin-dropdown-menu.html#L80)):*
<details><summary><i>Previously, tomivirkki (Tomi Virkki) wrote…</i></summary><blockquote>

Add listbox to this example
</blockquote></details>

Added.

---

*[vaadin-dropdown-menu.html, line 101 at r6](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-KwsIIXENzMqqPDeDuwr:-KwsjT97spU_-d0-_iJ7:bss4a2b) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/9f64c35df6e544f4d11665695bd0d6aa1ca48dc8/vaadin-dropdown-menu.html#L101)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

Why do we need this property?. It's used once in the same block
</blockquote></details>

Removed.

---

*[vaadin-dropdown-menu.html, line 131 at r6](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-Kwp09A-MW0QFEvFv54G:-KwsjWfRv5M2LGMJza1S:bzhelsx) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/9f64c35df6e544f4d11665695bd0d6aa1ca48dc8/vaadin-dropdown-menu.html#L131)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

1. we are avoiding to use ids because of avoiding conflicting id in shady mode, @tomivirkki  wdyt, should we remove the id here as well?
2. I don't think lastElementChild is the safest way to get the menu element, I would check for elements that has the mixin, or the localName
3. btw there is a overlayElement getter in this class, use it.
</blockquote></details>

2. Should we add a property to `ListMixin` like `_hasVaadinListMixin`?
3. Using now

---

*[vaadin-dropdown-menu.html, line 143 at r6](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-KwsGhKT9gbEn27VImHF:-KwsjiYZRe8kjXZ92-RI:b-a2ujfl) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/9f64c35df6e544f4d11665695bd0d6aa1ca48dc8/vaadin-dropdown-menu.html#L143)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

should use `e.key` instead
move boolean `evenFromSelected` at the beginning of the condition, checking booleans first make the condition perform better.
</blockquote></details>

Changed.

---

*[vaadin-dropdown-menu.html, line 146 at r6](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-KwsH6AT-76pCU2A6D8o:-KwsjjSLEqJ1g9_twvqS:b-a2ujfl) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/9f64c35df6e544f4d11665695bd0d6aa1ca48dc8/vaadin-dropdown-menu.html#L146)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

more concise `this.value === '' && this.value = e. ... ` or `this.value = this.value || ...`
</blockquote></details>

Changed.

---

*[vaadin-dropdown-menu.html, line 148 at r6](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-KwsHyoL7xPyUaJO9FXO:-KwsjkJtbKFv5cly_uO-:b-a2ujfl) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/9f64c35df6e544f4d11665695bd0d6aa1ca48dc8/vaadin-dropdown-menu.html#L148)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

`e.key`
</blockquote></details>

Changed.

---

*[vaadin-dropdown-menu.html, line 156 at r6](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-KwsIRgG3l3Gx951zREz:-Kwsjl_iEozwA_q5YYGU:b-896fix) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/9f64c35df6e544f4d11665695bd0d6aa1ca48dc8/vaadin-dropdown-menu.html#L156)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

Seems that it should be a `const _contentTemplate` instead of an instance property. It's not used anywhere else
</blockquote></details>

Done.

---

*[vaadin-dropdown-menu.html, line 174 at r6](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-KwsIiSWxd5polH7PjmM:-KwsjmpRk0MkAQBXXPNO:b-jb7y88) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/9f64c35df6e544f4d11665695bd0d6aa1ca48dc8/vaadin-dropdown-menu.html#L174)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

The same here. I guess menu could be computed once and leave it in a instance variable
</blockquote></details>

It is now used only in one block, I am not sure that we can compute it at any other point other than when overlay opens.

---

*[vaadin-dropdown-menu.html, line 187 at r6](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-KwpuYHZ2z8VG7xZHboE:-KwsjzWqGQTrO_leFtQL:b-6s83fw) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/9f64c35df6e544f4d11665695bd0d6aa1ca48dc8/vaadin-dropdown-menu.html#L187)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

what's the reason of adding/removing 
</blockquote></details>

I am adding it when menu exists and removing when it is not present. 

---

*[vaadin-dropdown-menu.html, line 199 at r6](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-KwsJ4O7fMD7l39jaOSU:-KwskB5FHo_6wobFwMo6:b343cc1) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/9f64c35df6e544f4d11665695bd0d6aa1ca48dc8/vaadin-dropdown-menu.html#L199)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

I feel like we should compute the menu element once when the light dom is available and add the events, I dont see the need of adding/removing events each time it's opened/closed
</blockquote></details>

I cannot find a proper moment when it should be computed, but I am free for suggestions and can change at any moment.

---

*[demo/dropdown-menu-basic-demos.html, line 11 at r6](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-KwsKfPT8BkYOv-UpaUj:-KwskM_yIpfk1qhS9xfH:b-a2ujfl) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/9f64c35df6e544f4d11665695bd0d6aa1ca48dc8/demo/dropdown-menu-basic-demos.html#L11)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

Don't we only support list-boxes as content? hence title should be just `Basic Usage` or `Usage`
</blockquote></details>

Changed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2)*
<!-- Sent from Reviewable.io -->

tomivirkki commented:
*[vaadin-dropdown-menu.html, line 215 at r6](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-KwsJY1twG-Aq9_2kc-i:-Kwslxdy2sKMEdzO1sRY:b62n0gt) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/9f64c35df6e544f4d11665695bd0d6aa1ca48dc8/vaadin-dropdown-menu.html#L215)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

@tomivirkki shouldn't we avoid ids ? 
</blockquote></details>

We sure should.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2)*
<!-- Sent from Reviewable.io -->

YuriyVaadin commented:
Review status: 0 of 13 files reviewed at latest revision, 18 unresolved discussions.

---

*[vaadin-dropdown-menu.html, line 215 at r6](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-KwsJY1twG-Aq9_2kc-i:-Kwtc2-eCtC26Wnp1Cll:b29277j) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/9f64c35df6e544f4d11665695bd0d6aa1ca48dc8/vaadin-dropdown-menu.html#L215)):*
<details><summary><i>Previously, tomivirkki (Tomi Virkki) wrote…</i></summary><blockquote>

We sure should.
</blockquote></details>

Avoided.

---

*[test/dropdown-menu-test.html, line 35 at r6](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-KwsL8FcceBCcwVkVKJ-:-Kwtc5XC3mbKX4ne_ahG:b-mshvfg) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/9f64c35df6e544f4d11665695bd0d6aa1ca48dc8/test/dropdown-menu-test.html#L35)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

we need more tests here, like testing selecting, focus code, etc
</blockquote></details>

Basic tests added.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2)*
<!-- Sent from Reviewable.io -->

tomivirkki commented:
*[vaadin-dropdown-menu.html, line 115 at r8](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-KwtjG_PLNlEKBljH8mh:-KwtjG_PLNlEKBljH8mi:bbsw1oc) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/2b424357be5bdea719df42050ad9a916a4c72f3f/vaadin-dropdown-menu.html#L115)):*
> ```HTML
> 
>             /**
>              * The value of the control, which is submitted with the form data.
> ```

Align API naming with https://vaadin.com/elements/vaadin-combo-box/html-api/elements/Vaadin.ComboBoxElement

Also, analysis.json needs an update

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2)*
<!-- Sent from Reviewable.io -->

tomivirkki commented:
Reviewed 10 of 15 files at r3, 1 of 2 files at r7, 3 of 3 files at r8, 2 of 2 files at r9.
Review status: all files reviewed at latest revision, 19 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2)*
<!-- Sent from Reviewable.io -->

manolo commented:
Reviewed 10 of 15 files at r3, 1 of 2 files at r7, 3 of 3 files at r8, 2 of 2 files at r10.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[vaadin-dropdown-menu.html, line 68 at r6](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-KwozmAz3V_nGChxxrL0:-Kx6mKuz_udB4hUvz2l2:b-w4yx3s) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/9f64c35df6e544f4d11665695bd0d6aa1ca48dc8/vaadin-dropdown-menu.html#L68)):*
<details><summary><i>Previously, YuriyVaadin (Yuriy Yevstihnyeyev) wrote…</i></summary><blockquote>

`name` and `disabled` were added. Should `validate` method be the same as in `radio-button-group` for example? https://github.com/vaadin/vaadin-radio-button/blob/master/vaadin-radio-button-group.html#L221
</blockquote></details>

yes it works to me

---

*[vaadin-dropdown-menu.html, line 174 at r6](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-KwsIiSWxd5polH7PjmM:-Kx6nBswygqNLSrgfjWR:b32wtd) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/9f64c35df6e544f4d11665695bd0d6aa1ca48dc8/vaadin-dropdown-menu.html#L174)):*
<details><summary><i>Previously, YuriyVaadin (Yuriy Yevstihnyeyev) wrote…</i></summary><blockquote>

It is now used only in one block, I am not sure that we can compute it at any other point other than when overlay opens.
</blockquote></details>

ok

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2)*
<!-- Sent from Reviewable.io -->

YuriyVaadin commented:
Review status: 10 of 13 files reviewed at latest revision, 9 unresolved discussions.

---

*[vaadin-dropdown-menu.html, line 68 at r6](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-KwozmAz3V_nGChxxrL0:-Kx7G1beqhBVncHiwSKM:bqmjwcz) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/9f64c35df6e544f4d11665695bd0d6aa1ca48dc8/vaadin-dropdown-menu.html#L68)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

yes it works to me
</blockquote></details>

Added.

---

*[vaadin-dropdown-menu.html, line 115 at r8](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-KwtjG_PLNlEKBljH8mh:-Kx7AwNqClYOXOxuL3Dm:b-896fix) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/2b424357be5bdea719df42050ad9a916a4c72f3f/vaadin-dropdown-menu.html#L115)):*
<details><summary><i>Previously, tomivirkki (Tomi Virkki) wrote…</i></summary><blockquote>

Align API naming with https://vaadin.com/elements/vaadin-combo-box/html-api/elements/Vaadin.ComboBoxElement

Also, analysis.json needs an update
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2)*
<!-- Sent from Reviewable.io -->

YuriyVaadin commented:
Review status: 10 of 13 files reviewed at latest revision, 9 unresolved discussions.

---

*[vaadin-dropdown-menu.html, line 124 at r6](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-Kwp-MnPpj_DbbCJblv8:-Kx7HJy6OiBZ-h_N7ZZ7:b-nbwbgq) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/9f64c35df6e544f4d11665695bd0d6aa1ca48dc8/vaadin-dropdown-menu.html#L124)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

Native select does not open with enter, but paper-drop-down does. I feel that we should fill native behavior but not sure @limonte @platosha  ?
</blockquote></details>

Can be changed in later PRs

---

*[vaadin-dropdown-menu.html, line 131 at r6](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-Kwp09A-MW0QFEvFv54G:-Kx7GxzJUyEdfL1_gMZX:b3067zy) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/9f64c35df6e544f4d11665695bd0d6aa1ca48dc8/vaadin-dropdown-menu.html#L131)):*
<details><summary><i>Previously, YuriyVaadin (Yuriy Yevstihnyeyev) wrote…</i></summary><blockquote>

2. Should we add a property to `ListMixin` like `_hasVaadinListMixin`?
3. Using now
</blockquote></details>

1. ids are avoid.
2. `lastElementChild` replaced with `querySelector`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2)*
<!-- Sent from Reviewable.io -->

manolo commented:
Review status: 10 of 13 files reviewed at latest revision, 12 unresolved discussions.

---

*[bower.json, line 38 at r11](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-Kx7ex9H6KyK9voTtIDt:-Kx7ex9H6KyK9voTtIDu:b-5ja26i) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/1a6d391e12e2f09c2f4ae1e2ccd00af1fcb6e645/bower.json#L38)):*
> ```JSON
>     "vaadin-demo-helpers": "vaadin/vaadin-demo-helpers",
>     "vaadin-text-field": "^1.1.0",
>     "vaadin-overlay": "^1.4.1",
> ```

text-field and vaadin-overlay shouldn't not be dev dependencies.
vaadin-list-box should be unless we support other implementations.

---

*[vaadin-dropdown-menu.html, line 233 at r11](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-Kx7gOy_vfDlQ6smOkGY:-Kx7gOyan5vkJdrJOUhk:bkakslh) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/1a6d391e12e2f09c2f4ae1e2ccd00af1fcb6e645/vaadin-dropdown-menu.html#L233)):*
> ```HTML
>           overlayElement.opened = opened;
> 
>           const menu = overlayElement.$.content.querySelector('vaadin-list-box');
> ```

This needs a rethinking, probably we could just get the firs element implementing the list-mixin

---

*[vaadin-dropdown-menu.html, line 235 at r11](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-Kx7iGJKc_aTCFE56xHT:-Kx7iGJKc_aTCFE56xHU:b5bauo3) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/1a6d391e12e2f09c2f4ae1e2ccd00af1fcb6e645/vaadin-dropdown-menu.html#L235)):*
> ```HTML
>           const menu = overlayElement.$.content.querySelector('vaadin-list-box');
>           const selectedChangedListener = (e) => {
>             const selectedItem = menu._navItems[e.detail.value];
> ```

This is a private API, needs at least to be marked as protected in the mixin

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2)*
<!-- Sent from Reviewable.io -->

limonte commented:
Reviewed 11 of 15 files at r3, 2 of 3 files at r6, 2 of 2 files at r7.
Review status: 10 of 13 files reviewed at latest revision, 12 unresolved discussions.

---

*[vaadin-dropdown-menu.html, line 124 at r6](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-Kwp-MnPpj_DbbCJblv8:-Kx7p1aiN1knWZgnO56e:bgrkvbe) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/9f64c35df6e544f4d11665695bd0d6aa1ca48dc8/vaadin-dropdown-menu.html#L124)):*
<details><summary><i>Previously, YuriyVaadin (Yuriy Yevstihnyeyev) wrote…</i></summary><blockquote>

Can be changed in later PRs
</blockquote></details>

I don't have any strong opinion on this. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2)*
<!-- Sent from Reviewable.io -->

YuriyVaadin commented:
Review status: 9 of 13 files reviewed at latest revision, 12 unresolved discussions.

---

*[bower.json, line 38 at r11](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-Kx7ex9H6KyK9voTtIDt:-Kx82aFwqYkHJ6Ot_sjn:b-896fix) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/1a6d391e12e2f09c2f4ae1e2ccd00af1fcb6e645/bower.json#L38)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

text-field and vaadin-overlay shouldn't not be dev dependencies.
vaadin-list-box should be unless we support other implementations.

</blockquote></details>

Done.

---

*[vaadin-dropdown-menu.html, line 233 at r11](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-Kx7gOy_vfDlQ6smOkGY:-Kx81kVwsXA5mR2OL9fR:b-tz9mwh) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/1a6d391e12e2f09c2f4ae1e2ccd00af1fcb6e645/vaadin-dropdown-menu.html#L233)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

This needs a rethinking, probably we could just get the firs element implementing the list-mixin
</blockquote></details>

Will be replaced with `Array.from(overlayElement.$.content.childNodes).filter(node => node._hasVaadinListMixin)[0]` once the new version of the `vaadin-list-mixin` will be released.

---

*[vaadin-dropdown-menu.html, line 235 at r11](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-Kx7iGJKc_aTCFE56xHT:-Kx82BhRuIL5UzcwzADB:bhodybu) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/1a6d391e12e2f09c2f4ae1e2ccd00af1fcb6e645/vaadin-dropdown-menu.html#L235)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

This is a private API, needs at least to be marked as protected in the mixin
</blockquote></details>

Should any changes be done to these part in `vaadin-dropdown-menu`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2)*
<!-- Sent from Reviewable.io -->

tomivirkki commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 15 files at r3, 3 of 3 files at r11, 1 of 1 files at r12.
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-:-Kx831uDqnkWdZXhDE2i:bnfp4nl)*
<!-- Sent from Reviewable.io -->

manolo commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks broke.

---

*[vaadin-dropdown-menu.html, line 187 at r6](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-KwpuYHZ2z8VG7xZHboE:-KxBvFqRp8F7lUvU6LrA:b-tfr1du) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/9f64c35df6e544f4d11665695bd0d6aa1ca48dc8/vaadin-dropdown-menu.html#L187)):*
<details><summary><i>Previously, YuriyVaadin (Yuriy Yevstihnyeyev) wrote…</i></summary><blockquote>

I am adding it when menu exists and removing when it is not present. 
</blockquote></details>

I meant that we can just add the listeners once and it's not necessary to remove continously, that simplifies the code.

---

*[vaadin-dropdown-menu.html, line 199 at r6](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-KwsJ4O7fMD7l39jaOSU:-KxBviAmYiQiCTLNs3_4:bkbvv1t) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/9f64c35df6e544f4d11665695bd0d6aa1ca48dc8/vaadin-dropdown-menu.html#L199)):*
<details><summary><i>Previously, YuriyVaadin (Yuriy Yevstihnyeyev) wrote…</i></summary><blockquote>

I cannot find a proper moment when it should be computed, but I am free for suggestions and can change at any moment.
</blockquote></details>

so we can do in a separated refactor, not a blocker

---

*[vaadin-dropdown-menu.html, line 233 at r11](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-Kx7gOy_vfDlQ6smOkGY:-KxBueZcedka0RkUrbrG:b-64zjoc) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/1a6d391e12e2f09c2f4ae1e2ccd00af1fcb6e645/vaadin-dropdown-menu.html#L233)):*
<details><summary><i>Previously, YuriyVaadin (Yuriy Yevstihnyeyev) wrote…</i></summary><blockquote>

Will be replaced with `Array.from(overlayElement.$.content.childNodes).filter(node => node._hasVaadinListMixin)[0]` once the new version of the `vaadin-list-mixin` will be released.
</blockquote></details>

Leave this with the query selector, since we are going to change the `vaadin-item` so it should be used `instanceof`, but you can also look for element that has `_navItems` since you are using that api below

---

*[vaadin-dropdown-menu.html, line 235 at r11](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-Kx7iGJKc_aTCFE56xHT:-KxBuuJWwYVb-_ldOXhn:bji2h5f) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/1a6d391e12e2f09c2f4ae1e2ccd00af1fcb6e645/vaadin-dropdown-menu.html#L235)):*
<details><summary><i>Previously, YuriyVaadin (Yuriy Yevstihnyeyev) wrote…</i></summary><blockquote>

Should any changes be done to these part in `vaadin-dropdown-menu`?
</blockquote></details>

we need to mark _navItems as protected so as it is not renamed or removed in the future

---

*[vaadin-dropdown-menu.html, line 168 at r12](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-KxBuUaBq_lnvhVgX-Ln:-KxBuUaBq_lnvhVgX-Lo:b-otf3s0) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/ba89f6f208a8681490d5338dbae32a5b57bdf7d2/vaadin-dropdown-menu.html#L168)):*
> ```HTML
>           this._inputElement = this.shadowRoot.querySelector('vaadin-text-field');
> 
>           this._inputElement.addEventListener('click', e => this._toggleOverlay());
> ```

Apart from toggling the overlay it needs to focus the listbox

---

*[vaadin-dropdown-menu.html, line 254 at r12](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-KxBrb__q3RSC2HUzcdd:-KxBrb__q3RSC2HUzcde:b-oyjsp5) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/ba89f6f208a8681490d5338dbae32a5b57bdf7d2/vaadin-dropdown-menu.html#L254)):*
> ```HTML
>             if (itemToFocus) {
>               itemToFocus.focus();
>             }
> ```

Replace this block with `menu.focus()` already supported in list-mixin 1.3.0

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2)*
<!-- Sent from Reviewable.io -->

manolo commented:
Review status: 11 of 13 files reviewed at latest revision, 4 unresolved discussions.

---

*[vaadin-dropdown-menu.html, line 168 at r12](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-KxBuUaBq_lnvhVgX-Ln:-KxCHlcHjdpg8OUyyc_Q:bb6un1j) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/ba89f6f208a8681490d5338dbae32a5b57bdf7d2/vaadin-dropdown-menu.html#L168)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

Apart from toggling the overlay it needs to focus the listbox
</blockquote></details>

Done

---

*[vaadin-dropdown-menu.html, line 254 at r12](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-KxBrb__q3RSC2HUzcdd:-KxCHohggqjlMaI8Ja4v:bb6un1j) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/ba89f6f208a8681490d5338dbae32a5b57bdf7d2/vaadin-dropdown-menu.html#L254)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

Replace this block with `menu.focus()` already supported in list-mixin 1.3.0
</blockquote></details>

Done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2)*
<!-- Sent from Reviewable.io -->

manolo commented:
Review status: 11 of 13 files reviewed at latest revision, 4 unresolved discussions.

---

*[vaadin-dropdown-menu.html, line 187 at r6](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-KwpuYHZ2z8VG7xZHboE:-KxCMrUS43BI3c1Ef6hb:bb6un1j) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/9f64c35df6e544f4d11665695bd0d6aa1ca48dc8/vaadin-dropdown-menu.html#L187)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

I meant that we can just add the listeners once and it's not necessary to remove continously, that simplifies the code.
</blockquote></details>

Done

---

*[vaadin-dropdown-menu.html, line 235 at r11](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-Kx7iGJKc_aTCFE56xHT:-KxCMvyJvphsK57xZ4zf:b-lzeqa0) ([raw file](https://github.com/vaadin/vaadin-dropdown-menu/blob/1a6d391e12e2f09c2f4ae1e2ccd00af1fcb6e645/vaadin-dropdown-menu.html#L235)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

we need to mark _navItems as protected so as it is not renamed or removed in the future
</blockquote></details>

We would do later, not a blocker for this PR

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2)*
<!-- Sent from Reviewable.io -->

manolo commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 11 of 13 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-dropdown-menu/2#-:-KxCTiuEOEqEaN7UH7lb:bnfp4nl)*
<!-- Sent from Reviewable.io -->
